Vostok.Logging.File/Configuration/SafeSettingsCache.cs

```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Vostok.Logging.File.Configuration
+{
+    internal class SafeSettingsCache
+    {
+        private static readonly object CooldownGuard = new object();
+
+        private readonly SafeSettingsProvider provider;
+        private readonly TimeSpan ttl = TimeSpan.FromSeconds(1);
+        private readonly bool enabled;
+
+        private volatile object updateCooldown;
+        private volatile Task updateCacheTask = Task.CompletedTask;
+        private volatile FileLogSettings currentSettings;
+
+        public SafeSettingsCache(Func<FileLogSettings> provider)
+        {
+            this.provider = new SafeSettingsProvider(provider);
+
+            var settings = Get();
+
+            enabled = settings.EnableFileLogSettingsCache;
+
+            if (enabled)
+                currentSettings = settings;
+        }
+
+        public FileLogSettings Get()
+        {
+            if (!enabled)
+                return provider.Get();
+
+            ScheduleRefresh();
+
+            return currentSettings;
+        }
+
+        public void ForceRefresh() => Interlocked.Exchange(ref currentSettings, provider.UnsafeGet());
+
+        private void ScheduleRefresh()
+        {
+            if (updateCooldown == null && Interlocked.CompareExchange(ref updateCooldown, CooldownGuard, null) == null)
+            {
+                updateCacheTask = Task.Run(
+                    () =>
+                    {
+                        var oldSettings = currentSettings;
+                        var newSettings = provider.Get();
+                        Interlocked.CompareExchange(ref currentSettings, newSettings, oldSettings);
+                    });
+
+                updateCacheTask
+                   .ContinueWith(_ => Task.Delay(ttl))
+                   .Unwrap()
+                   .ContinueWith(_ => updateCooldown = null);
```

Comment:
Shouldn't we reset `updateCooldown` with `Interlocked.Exchange` here?

Reply:
Looks like only one thread at a time is able to change this value, so I don't see any data races here actually.

CompareExchange attempts to change iff `updateCooldown` is null, but it isn't at a time of this reset, so I think that there is no way we can lose this assignment,